File: Android/app/src/main/java/com/edu/primary/repository/AIRepository.java
package com.edu.primary.repository;

import android.content.Context;
import android.content.SharedPreferences;
import com.edu.primary.constants.AppConstants;
import com.edu.primary.database.AppDatabase;
import com.edu.primary.database.entity.ChatMessageEntity;
import com.edu.primary.network.ApiClient;
import com.edu.primary.network.DeepseekApiService;
import com.edu.primary.network.model.DeepseekRequest;
import com.edu.primary.network.model.DeepseekResponse;
import io.reactivex.Single;
import io.reactivex.schedulers.Schedulers;

import java.util.ArrayList;
import java.util.List;

public class AIRepository {

    private final AppDatabase database;
    private final SharedPreferences prefs;
    private final DeepseekApiService apiService;
    private final Context appContext; // 使用ApplicationContext防止内存泄漏

    /**
     * 构造函数
     * 使用ApplicationContext防止内存泄漏
     */
    public AIRepository(Context context) {
        // 使用ApplicationContext防止内存泄漏
        this.appContext = context.getApplicationContext();
        database = AppDatabase.getInstance(appContext);
        prefs = appContext.getSharedPreferences(AppConstants.PREFS_AI, Context.MODE_PRIVATE);
        apiService = ApiClient.getDeepseekApiService();
    }

    public void setApiKey(String apiKey) {
        prefs.edit().putString(AppConstants.KEY_API_KEY, apiKey).apply();
    }

    public String getApiKey() {
        return prefs.getString(AppConstants.KEY_API_KEY, "");
    }

    public Single<String> sendMessage(long userId, String message) {
        return Single.fromCallable(() -> {
            ChatMessageEntity userMessage = new ChatMessageEntity(userId, AppConstants.ROLE_USER, message);
            database.chatMessageDao().insertMessage(userMessage);

            List<ChatMessageEntity> history = database.chatMessageDao().getMessagesByUser(userId);
            
            if (history.size() > AppConstants.MAX_HISTORY_MESSAGES) {
                List<ChatMessageEntity> recentHistory = new ArrayList<>();
                for (int i = history.size() - AppConstants.MAX_HISTORY_MESSAGES; i < history.size(); i++) {
                    recentHistory.add(history.get(i));
                }
                history = recentHistory;
            }

            DeepseekRequest request = new DeepseekRequest();
            List<DeepseekRequest.Message> messages = new ArrayList<>();
            
            messages.add(new DeepseekRequest.Message(AppConstants.ROLE_SYSTEM, 
                appContext.getString(com.edu.primary.R.string.ai_system_prompt)));
            
            for (ChatMessageEntity msg : history) {
                messages.add(new DeepseekRequest.Message(msg.role, msg.content));
            }
            
            request.setMessages(messages);

            String apiKey = getApiKey();
            if (apiKey.isEmpty()) {
                throw new Exception(AppConstants.ERROR_API_KEY_NOT_SET);
            }

            DeepseekResponse response = apiService.chat(
                AppConstants.BEARER_PREFIX + apiKey,
                AppConstants.CONTENT_TYPE_JSON,
                request
            ).blockingGet();

            String assistantMessage = response.getContent();
            
            ChatMessageEntity assistantMsg = new ChatMessageEntity(userId, AppConstants.ROLE_ASSISTANT, assistantMessage);
            database.chatMessageDao().insertMessage(assistantMsg);

            return assistantMessage;
        }).subscribeOn(Schedulers.io());
    }

    public Single<List<ChatMessageEntity>> getChatHistory(long userId) {
        return Single.fromCallable(() ->
            database.chatMessageDao().getMessagesByUser(userId)
        ).subscribeOn(Schedulers.io());
    }

    /**
     * 清空聊天历史
     * 使用RxJava替代new Thread
     */
    public io.reactivex.Completable clearChatHistory(long userId) {
        return io.reactivex.Completable.fromAction(() ->
            database.chatMessageDao().clearMessages(userId)
        ).subscribeOn(io.reactivex.schedulers.Schedulers.io());
    }
}

File: Android-1/app/src/main/java/com/edu/primary/data/repository/AIRepository.kt
package com.edu.primary.data.repository

import android.content.Context
import com.edu.primary.ai.RuleBasedTextGenerator
import com.edu.primary.data.local.AppDatabase
import com.edu.primary.data.local.entity.ChatMessageEntity
import com.edu.primary.util.AppConstants
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext

/**
 * AI 助手数据层：聊天记录持久化，回复生成优先本地 TFLite（可扩展），无模型时使用 [RuleBasedTextGenerator] 规则回退。
 *
 * 参考 aiAssistant 的 LocalAiEngine + RuleBasedTextGenerator。消息存入 [ChatMessageDao]，历史条数受 [AppConstants.MAX_HISTORY_MESSAGES] 限制。
 *
 * @param context 用于获取 ApplicationContext 与 SharedPreferences
 */
class AIRepository(context: Context) {
    private val appContext = context.applicationContext
    private val db = AppDatabase.getInstance(appContext)
    private val prefs = appContext.getSharedPreferences(AppConstants.PREFS_AI, Context.MODE_PRIVATE)

    /** 保存 API Key（若将来接入云端接口时使用）。 */
    fun setApiKey(key: String) {
        prefs.edit().putString(AppConstants.KEY_API_KEY, key).apply()
    }
    /** 读取已保存的 API Key。 */
    fun getApiKey(): String = prefs.getString(AppConstants.KEY_API_KEY, "").orEmpty()

    /**
     * 发送用户消息并生成助手回复。先持久化用户消息，再根据历史调用 [RuleBasedTextGenerator.chat] 生成回复并持久化。
     * @param userId 当前用户 ID
     * @param content 用户输入内容
     * @return [Result.success] 为助手回复文本，[Result.failure] 为异常
     */
    suspend fun sendMessage(userId: Long, content: String): Result<String> = withContext(Dispatchers.IO) {
        val userMsg = ChatMessageEntity(userId = userId, role = AppConstants.ROLE_USER, content = content)
        db.chatMessageDao().insert(userMsg)
        val history = db.chatMessageDao().getByUser(userId)
        val trimmed = if (history.size > AppConstants.MAX_HISTORY_MESSAGES)
            history.takeLast(AppConstants.MAX_HISTORY_MESSAGES) else history
        val reply = RuleBasedTextGenerator.chat(content, trimmed.size)
        val assistantMsg = ChatMessageEntity(userId = userId, role = AppConstants.ROLE_ASSISTANT, content = reply)
        db.chatMessageDao().insert(assistantMsg)
        Result.success(reply)
    }

    /** 获取指定用户的聊天历史，按时间升序。 */
    suspend fun getChatHistory(userId: Long): List<ChatMessageEntity> = withContext(Dispatchers.IO) {
        db.chatMessageDao().getByUser(userId)
    }

    /** 清空指定用户的聊天记录。 */
    suspend fun clearChatHistory(userId: Long) = withContext(Dispatchers.IO) {
        db.chatMessageDao().clearByUser(userId)
    }
}
